source/opt/eliminate_dead_output_stores_pass.cpp

```diff
@@ -0,0 +1,418 @@
+        ocnt++;
+        return true;
+      }
+      // If any non-constant index, mark the entire current object and return.
+      auto idx_inst = def_use_mgr->GetDef(*opnd);
+      if (idx_inst->opcode() != SpvOpConstant)
+        return false;
+      // If current type is struct, look for location decoration on member and
+      // reset offset if found.
+      auto index = idx_inst->GetSingleWordInOperand(0);
+      auto str_type = (*curr_type)->AsStruct();
+      if (str_type) {
+        uint32_t loc = 0;
+        auto str_type_id = type_mgr->GetId(str_type);
+        bool no_mem_loc = deco_mgr->WhileEachDecoration(
+          str_type_id, SpvDecorationLocation,
+          [&loc, index, no_loc](const Instruction& deco) {
+          assert(deco.opcode() == SpvOpMemberDecorate && "unexpected decoration");
+          if (deco.GetSingleWordInOperand(kOpDecorateMemberMemberInIdx) == index) {
+            loc = deco.GetSingleWordInOperand(kOpDecorateMemberLocationInIdx);
+            *no_loc = false;
+            return false;
+          }
+          return true;
+        });
+        if (!no_mem_loc) {
+          *offset = loc;
+          *curr_type = GetComponentType(index, *curr_type);
+          ocnt++;
+          return true;
+        }
+      }
+
+      // Update offset and current type based on constant index.
+      *offset += GetLocOffset(index, *curr_type);
+      *curr_type = GetComponentType(index, *curr_type);
+    }
+    ocnt++;
+    return true;
+  });
+}
+
+void EliminateDeadOutputStoresPass::MarkRefLive(const Instruction* ref,
+    Instruction* var) {
+  analysis::DefUseManager* def_use_mgr = context()->get_def_use_mgr();
+  analysis::TypeManager* type_mgr = context()->get_type_mgr();
+  analysis::DecorationManager* deco_mgr = context()->get_decoration_mgr();
+  // Find variable location if present.
+  uint32_t loc = 0;
+  auto var_id = var->result_id();
+  bool no_loc = deco_mgr->WhileEachDecoration(
+    var_id, SpvDecorationLocation,
+    [&loc](const Instruction& deco) {
+    assert(deco.opcode() == SpvOpDecorate && "unexpected decoration");
+    loc = deco.GetSingleWordInOperand(kDecorationLocationInIdx);
+    return false;
+  });
+  // If use is a load, mark all locations of var
+  auto ptr_type = type_mgr->GetType(var->type_id())->AsPointer();
+  assert(ptr_type && "unexpected var type");
+  auto var_type = ptr_type->pointee_type();
+  if (ref->opcode() == SpvOpLoad) {
+    assert(!no_loc && "missing input variable location");
+    MarkLocsLive(loc, GetLocSize(var_type));
+    return;
+  }
+  // Mark just those locations indicated by access chain
+  assert((ref->opcode() == SpvOpAccessChain ||
+      ref->opcode() == SpvOpInBoundsAccessChain) &&
+      "unexpected use of input variable");
+  bool all_loads = def_use_mgr->WhileEachUser(ref, [](Instruction* use) {
+    if (use->opcode() != SpvOpLoad)
+      return false;
+    return true;
+  });
+  USE_ASSERT(all_loads && "unexpected non-load use of input access chain");
```

Comment:
> Why can't you have an access chain feeding an access chain feeding a load? Is it something particular about input variables?

It is theoretically possible, but I have rarely seen it, and never on input and output variables. I didn't want to spend time and complexity on something that would in all likelihood never be seen. 

In actuality, this analysis is conservative in the case of a nested access chain. That is, it will give a correct, if less than optimal answer.

Still, I wanted to catch it in case it did occur, especially during development.

> This will still traverse each user even if the assert is disabled. It might be best to just do an "assert" that calls a function AllUsesAreLoads(ref).

OK.

Reply:
In the end I decided to just remove this code.